Content.Server/Silicon/WeldingHealable/WeldingHealableSystem.cs

```diff
@@ -1,121 +1,102 @@
-            }
-
-            var str = Loc.GetString("comp-repairable-repair",
-                ("target", uid),
-                ("tool", args.Used!));
-            _popup.PopupEntity(str, uid, args.User);
-
-
-            if (args.Used.HasValue)
-            {
-                args.Handled = _toolSystem.UseTool(args.Used.Value, args.User, uid, args.Delay, component.QualityNeeded, new SiliconRepairFinishedEvent
-                {
-                    Delay = args.Delay
-                });
-            }
-        }
-
-
+namespace Content.Server.Silicon.WeldingHealable;
 
-        private async void Repair(EntityUid uid, WeldingHealableComponent healableComponent, InteractUsingEvent args)
-        {
-            if (args.Handled
-                || !EntityManager.TryGetComponent(args.Used, out WeldingHealingComponent? component)
-                || !EntityManager.TryGetComponent(args.Target, out DamageableComponent? damageable)
-                || damageable.DamageContainerID != null
-                && !component.DamageContainers.Contains(damageable.DamageContainerID)
-                || !HasDamage(damageable, component)
-                || !_toolSystem.HasQuality(args.Used, component.QualityNeeded))
-                return;
+public sealed class WeldingHealableSystem : SharedWeldingHealableSystem
+{
+    [Dependency] private readonly SharedToolSystem _toolSystem = default!;
+    [Dependency] private readonly DamageableSystem _damageableSystem = default!;
+    [Dependency] private readonly SharedPopupSystem _popup = default!;
+    [Dependency] private readonly SharedSolutionContainerSystem _solutionContainer = default!;
 
-            float delay = component.DoAfterDelay;
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<WeldingHealableComponent, InteractUsingEvent>(Repair);
+        SubscribeLocalEvent<WeldingHealableComponent, SiliconRepairFinishedEvent>(OnRepairFinished);
+    }
 
-            // Add a penalty to how long it takes if the user is repairing itself
-            if (args.User == args.Target)
+    private void OnRepairFinished(EntityUid uid, WeldingHealableComponent healableComponent, SiliconRepairFinishedEvent args)
+    {
+        if (args.Cancelled || args.Used == null
+            || !TryComp<DamageableComponent>(args.Target, out var damageable)
+            || !TryComp<WeldingHealingComponent>(args.Used, out var component)
+            || damageable.DamageContainerID is null
+            || !component.DamageContainers.Contains(damageable.DamageContainerID)
+            || !HasDamage(damageable, component)
+            || !TryComp<WelderComponent>(args.Used, out var welder)
+            || !TryComp<SolutionContainerManagerComponent>(args.Used, out var solutionContainer)
+            || !_solutionContainer.ResolveSolution(((EntityUid) args.Used, solutionContainer), welder.FuelSolutionName, ref welder.FuelSolution))
```

Comment:
it should be probably better ordered. `args.Used` checks first, then `args.Target`.